File: tools/check-connection-public-contract-v2.mjs
#!/usr/bin/env node
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import ts from "typescript";

const root = resolve(import.meta.dirname, "..");
const sourcePath = resolve(root, "src", "panel-framework", "connection", "routerAddress.ts");
const source = readFileSync(sourcePath, "utf8");
const { outputText, diagnostics } = ts.transpileModule(source, {
  compilerOptions: {
    module: ts.ModuleKind.ESNext,
    target: ts.ScriptTarget.ES2022,
  },
  fileName: sourcePath,
  reportDiagnostics: true,
});

if (diagnostics?.length) {
  throw new Error(ts.formatDiagnosticsWithColorAndContext(diagnostics, {
    getCanonicalFileName: (fileName) => fileName,
    getCurrentDirectory: () => root,
    getNewLine: () => "\n",
  }));
}

const { validateRouterAddress } = await import(`data:text/javascript;base64,${Buffer.from(outputText).toString("base64")}`);

const allowed = [
  "router.lan",
  "xn--bcher-kva.example",
  "bücher.example",
  "0.0.0.0",
  "255.255.255.255",
  "::1",
  "2001:db8::1",
  "[2001:db8::1]",
  "::ffff:192.0.2.1",
];
const rejected = [
  "256.0.0.1",
  "1.2.3.999",
  "router.lan:8728",
  "[2001:db8::1]:8728",
  "https://router.lan",
  "admin@router.lan",
  "router.lan?debug=1",
  "router.lan#fragment",
  "router.lan/rest",
];

for (const value of allowed) {
  if (validateRouterAddress(value) !== "") throw new Error(`expected address to be allowed: ${value}`);
}
for (const value of rejected) {
  if (validateRouterAddress(value) === "") throw new Error(`expected address to be rejected: ${value}`);
}

console.log("connection public contract v2 frontend: pass");

File: tools/check-connection-public-contract-v2.py
#!/usr/bin/env python3
"""Focused backend regression checks for the password-free RouterOS connection contract."""

from __future__ import annotations

import json
import sys
import tempfile
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))

from panel_backend.config_store import RouterProfileStore, RouterProfileStoreCorruptError
from panel_backend.router_transport import normalize_router_host


def expect_value_error(value: str) -> None:
    try:
        normalize_router_host(value)
    except ValueError:
        return
    raise AssertionError(f"RouterOS address unexpectedly accepted: {value!r}")


def assert_address_contract() -> None:
    assert normalize_router_host("192.168.88.1") == "192.168.88.1"
    assert normalize_router_host("router.lan") == "router.lan"
    assert normalize_router_host("bücher.example") == "xn--bcher-kva.example"
    assert normalize_router_host("[2001:db8::1]") == "2001:db8::1"
    for value in (
        "https://router.lan",
        "admin@router.lan",
        "router.lan?debug=1",
        "router.lan#fragment",
        "router.lan/rest",
        "router.lan:8728",
        "[2001:db8::1]:8728",
        "256.0.0.1",
    ):
        expect_value_error(value)


def assert_corrupt_store_never_becomes_an_empty_profile_list() -> None:
    with tempfile.TemporaryDirectory() as tempdir:
        path = Path(tempdir) / "router-profiles.json"
        store = RouterProfileStore(path)

        path.write_text(json.dumps({"entries": [{}]}), encoding="utf-8")
        try:
            store.load_unlocked()
        except RouterProfileStoreCorruptError:
            pass
        else:
            raise AssertionError("invalid saved profile was silently treated as an empty list")

        try:
            store.persist_unlocked([{}])
        except RouterProfileStoreCorruptError:
            pass
        else:
            raise AssertionError("invalid profile write was silently discarded")


def main() -> None:
    assert_address_contract()
    assert_corrupt_store_never_becomes_an_empty_profile_list()
    print("connection public contract v2: pass")


if __name__ == "__main__":
    main()
